fifo.c

```diff
@@ -2,82 +2,317 @@
+	    populate_buf((char*)buf, size);         // populate the message buf
+	    // printf("buf1=%s\n", (char*) buf+SIZE_OF_USED_BUFFER); 
+            if (write(fd1, buf, size) != size) {       // send message 1st time to parent
+                perror("first write");
+                unlink("./fifo-ipc1");
+                unlink("./fifo-ipc2");
+                return 1;
+            }
+            n = read(fd2, buf, size);      // receive message back from parent
+            if (n == -1) {
+                perror("read second");
+                unlink("./fifo-ipc1");
+                unlink("./fifo-ipc2");
                 return 1;
             }
+            // printf("buf2=%s\n", (char*)buf+SIZE_OF_USED_BUFFER); 
+    
+            gettimeofday(&tv_post_read, NULL);   // capture round trip timestamp
+            //printf("%.0f buf->tv_time_sent %d   second read \n", (double) buf->tv_time_sent.tv_usec, buf->seq); 
+            time_elapsed = time_diff(buf->tv_time_sent, tv_post_read);   // get latency
+            lat[msgtot].start_time = ((double) buf->tv_time_sent.tv_sec*1000000 + (double)buf->tv_time_sent.tv_usec);
+            latencies[msgtot] = time_elapsed;
+            lat[msgtot].end_time = ((double) tv_post_read.tv_sec*1000000 + (double)tv_post_read.tv_usec);
+            tot_elapsed += time_elapsed;
+
+    	    msgtot++;
+         
+            if  (msgtot == new_size)  {
+                    new_size += ARRAY_BOOST;
+                    double *latencies_tmp = realloc(latencies, new_size * sizeof(double));
+                    if (latencies_tmp == NULL)  {
+                        printf("Size increase of the latencies array failed.\n");
+                        return 1;
+                    }
+
+                    latencies_arr *lat_tmp = realloc(lat, new_size * sizeof(latencies_arr));
+                    if (lat_tmp == NULL)  {
+                        printf("Size increase of the lat structure array failed.\n");
+                        return 1;
+                    }
+                    latencies = latencies_tmp;
+                    lat = lat_tmp;
+             }
         }
+
         gettimeofday(&end, NULL);
 
         double tm = getdetlatimeofday(&begin, &end);
-        printf("%.0fMB/s %.0fmsg/s\n",
-            count * size * 1.0 / (tm * 1024 * 1024),
-            count * 1.0 / tm);
+
+        printf("%.0fMB/s\n", (msgtot*2) * size * 1.0 / (tm * 1024 * 1024));  // we double msgtot as the message went to parent and back 
+        printf("%.0fmsg/s\n", (msgtot*2) * 1.0 / tm);
+
+        avg_elapsed = tot_elapsed/msgtot;
+        ninetyninth = (msgtot*0.99)-1;
+        ninetyfifth = (msgtot*0.95)-1;
+
+        fprintf(outfile_latencies, "[\n");
+        for (ctr = 0; ctr < msgtot-1; ctr++) {
+            fprintf(outfile_latencies, "  {\"start-time\": %.0f, \"latency\": %.0f, \"end-time\": %.0f},\n", lat[ctr].start_time, latencies[ctr], lat[ctr].end_time);
+        }
```

Comment:
We may want to consider changing the schema of the latency json files. They can have an extremely large number of entries, and repeating the key names for each list item contributes significantly to the file size. Perhaps:
```
[
  ["start-time", "end-time", "latency"],
  [<value>, <value>, <value>],
  ...
]
```
or
```
[
  "start_time": [<value>, <value>, <value>, ...],
  "end_time": [<value>, <value>, <value>, ...],
  "latency": [<value>, <value>, <value>, ...],
]
```